finality-aleph/src/testing/network.rs

```diff
@@ -297,11 +298,178 @@ impl TestData {
+        while let Some(message) = self.network.send_message.try_next().await {
+            if let (NetworkData::Data(data, session_id), peer_id, protocol) = message {
+                panic!("No Data messages should be sent during cleanup. All data messages should be handled before.\
+                 Got: {:?} in {:?} to {:?} with protocol {:?}", data, session_id, peer_id, protocol);
+            }
+        }
         self.network.close_channels().await;
     }
 }
 
+#[tokio::test]
+async fn test_sends_discovery_message() {
+    let session_id = 43;
+    let mut test_data = prepare_one_session_test_data().await;
+    let connected_peer_id = test_data.authorities[1].peer_id();
+    test_data.connect_identity_to_network(connected_peer_id, Protocol::Generic);
+    let mut data_network = test_data.start_validator_session(0, session_id).await;
+    let handler = test_data.get_session_handler(0, session_id).await;
+
+    for _ in 0..5 {
+        if let Some((
+            NetworkData::Meta(DiscoveryMessage::AuthenticationBroadcast(auth_data)),
+            peer_id,
+            protocol,
+        )) = timeout(DEFAULT_TIMEOUT, test_data.network.send_message.next())
+            .await
+            .ok()
+            .flatten()
+        {
+            assert_eq!(peer_id, connected_peer_id);
+            assert_eq!(protocol, Protocol::Generic.name());
+            assert_eq!(auth_data, handler.authentication().unwrap());
+        } else {
+            panic!("Should broadcast authentication");
+        }
+    }
+
+    test_data.cleanup().await;
+    assert_eq!(
+        timeout(DEFAULT_TIMEOUT, data_network.next()).await,
+        Ok(None)
+    );
+}
+
+#[tokio::test]
+async fn test_sends_authentication_on_receiving_broadcast() {
+    let session_id = 43;
+    let mut test_data = prepare_one_session_test_data().await;
+    let mut data_network = test_data.start_validator_session(0, session_id).await;
+    let handler = test_data.get_session_handler(0, session_id).await;
+    let sending_peer_handler = test_data.get_session_handler(1, session_id).await;
+    let sending_peer = test_data.authorities[1].clone();
+    test_data.connect_identity_to_network(sending_peer.peer_id(), Protocol::Generic);
+
+    test_data.network.emit_event(Event::NotificationsReceived {
+        remote: sending_peer.peer_id().into(),
+        messages: vec![(
+            Protocol::Generic.name(),
+            NetworkData::<MockData>::Meta(DiscoveryMessage::AuthenticationBroadcast(
+                sending_peer_handler.authentication().unwrap(),
+            ))
+            .encode()
+            .into(),
+        )],
+    });
+
+    assert_eq!(
+        timeout(DEFAULT_TIMEOUT, test_data.network.add_reserved.next())
+            .await
+            .ok()
+            .flatten()
+            .expect("Should add reserved nodes"),
+        (
+            sending_peer.addresses().into_iter().collect(),
+            Protocol::Validator.name()
+        ),
+    );
+
+    loop {
```

Comment:
I guess we could add an upper bound on how many other messages it receives. I think something big like 20 other messages should be fine. Also I guess I could add an upper bound like that in other places where loop might run infinitely. WDYT?

Reply:
Ok, I have a better idea. I created and async function that will return only upon receiving wanted data, so in case we don't receive it the timeout will end.